### DISCUSSION TITLE
How to stream to pipewire-pulse with WebRTC using CLI tools? WHEP cURL wget ffmpeg

### DISCUSSION THREAD

**Discussion by iconoclasthero** [bluenviron/mediamtx#3645] [Questions]
### Question

This is closely related to my question about streaming mpd to mediamtx (url needed).
I have a working mpd to ffmpeg to mediamtx to chrome over webrtc stream working.
I want pulseaudio to receive that stream more directly and without need for GUI tools.

Since WebRTC cannot use cURL or wget directly, it appears that I could use WHEP to either capture with them and send to ffmpeg or just capture directly with ffmpeg.  Once ffmpeg has it, I believe sending it to pulse is trivial and ffplay is probably acceptable.  There may be something in pipewire or pipewire-pulse that received webrtc/whep directly and that works too.

Anyway, as someone else noted there aren't examples for WHEP and I'm lost setting this up.

I assume that if MediaMTX has a stream published from wherever, then it's MediaMTX's stream at that point (i.e., nothing more needed from the publisher).

So then I'm looking for help making sure that MediaMTX can stream this (no authentication) via WebRTC/WHEP and the corresponding information necessary to use some tool to accomplish the titular goal?

**Comment by aler9** [comment]
> There may be something in pipewire or pipewire-pulse that received webrtc/whep directly and that works too.

You can stream WebRTC/WHEP into pipewire by using GStreamer, this is also one of the suggested methods for ingesting network streams into pipewire:

```
gst-launch-1.0 whepsrc whep-endpoint="http://127.0.0.1:8889/stream/whep" use-link-headers=true \
audio-caps="application/x-rtp,media=audio,encoding-name=OPUS,payload=111,clock-rate=48000,encoding-params=(string)2" \
! rtpopusdepay ! decodebin ! pipewiresink
```

Remember to edit `audio-caps` and `video-caps` according to the stream codecs. The README contains sample pipelines for handling video, audio and video+audio with whepsrc.
